packages/core/saved-objects/core-saved-objects-migration-server-internal/src/document_migrator/validate_migrations.ts

```diff
@@ -8,114 +8,180 @@
+    const switchToModelVersionAt = Semver.parse(type.switchToModelVersionAt);
+    if (!switchToModelVersionAt) {
       throw new Error(
-        `Invalid convertToMultiNamespaceTypeVersion for type ${type}. Value '${convertToMultiNamespaceTypeVersion}' cannot be any other than '${convertVersion}'.`
-      );
-    } else if (Semver.gt(convertToMultiNamespaceTypeVersion, kibanaVersion)) {
-      throw new Error(
-        `Invalid convertToMultiNamespaceTypeVersion for type ${type}. Value '${convertToMultiNamespaceTypeVersion}' cannot be greater than the current Kibana version '${kibanaVersion}'.`
-      );
-    } else if (Semver.patch(convertToMultiNamespaceTypeVersion)) {
-      throw new Error(
-        `Invalid convertToMultiNamespaceTypeVersion for type ${type}. Value '${convertToMultiNamespaceTypeVersion}' cannot be used on a patch version (must be like 'x.y.0').`
+        `Type ${type.name}: invalid version specified for switchToModelVersionAt: ${type.switchToModelVersionAt}`
       );
     }
+    if (switchToModelVersionAt.patch !== 0) {
+      throw new Error(`Type ${type.name}: can't use a patch version for switchToModelVersionAt`);
+    }
   }
 
-  registry.getAllTypes().forEach((type) => {
-    const { name, migrations, convertToMultiNamespaceTypeVersion, namespaceType } = type;
-    if (migrations) {
-      assertObjectOrFunction(
-        type.migrations,
-        `Migration for type ${name} should be an object or a function returning an object like { '2.0.0': (doc) => doc }.`
-      );
-    }
-    if (convertToMultiNamespaceTypeVersion) {
-      // CHECKPOINT 1
-      assertValidConvertToMultiNamespaceType(
-        namespaceType,
-        convertToMultiNamespaceTypeVersion,
-        name
-      );
-    }
-  });
-}
+  if (type.migrations) {
+    assertObjectOrFunction(
+      type.migrations,
+      `Migration for type ${type.name} should be an object or a function returning an object like { '2.0.0': (doc) => doc }.`
+    );
 
-export function validateMigrationsMapObject(
-  name: string,
-  kibanaVersion: string,
-  migrationsMap?: SavedObjectMigrationMap
-) {
-  function assertObject(obj: any, prefix: string) {
-    if (!obj || typeof obj !== 'object') {
-      throw new Error(`${prefix} Got ${obj}.`);
-    }
+    const migrationMap =
+      typeof type.migrations === 'function' ? type.migrations() : type.migrations ?? {};
+
+    assertObject(
+      migrationMap,
+      `Migrations map for type ${type.name} should be an object like { '2.0.0': (doc) => doc }.`
+    );
+
+    Object.entries(migrationMap).forEach(([version, fn]) => {
+      assertValidSemver(kibanaVersion, version, type.name);
+      assertValidTransform(fn, version, type.name);
+      if (type.switchToModelVersionAt && Semver.gte(version, type.switchToModelVersionAt)) {
+        throw new Error(
+          `Migration for type ${type.name} for version ${version} registered after switchToModelVersionAt (${type.switchToModelVersionAt})`
```

Comment:
aside: this will catch copy-paste errors as an early warning to folks adding new migrations and let them know what's wrong. Thanks!